### PR TITLE
Add request headers to request.input

### LINKED ISSUE #55: add request headers to request.input
see https://github.com/kuzzleio/kuzzle-proxy/issues/76

### PR DESCRIPTION
**Pull request by ballinette**
This enable Kuzzle to manage request's HTTP headers (for both HTTP protocol as well as over-HTTP ones, like Websocket and SocketIO), and, for example, setup permissions restrictions according to some headers values.

fix #55 

**Boy-scout:**
Fix linter settings and fix eslint errors

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,6 +62,7 @@ This constructor is used to transform an [API request](http://kuzzle.io/api-refe
 | `connectionId` | `string` | Passed to [RequestContext](#modelsrequestcontext) constructor |
 | `error` | `KuzzleError` or `Error` | Invokes [setError](#seterrorerror) at initialization |
 | `protocol` | `string` | Passed to [RequestContext](#modelsrequestcontext) constructor |
+| `headers` | `object` | Passed to [RequestContext](#modelsrequestcontext) constructor |
```
Comment (scottinet) [inline README.md]: passed to RequestInput

File: README.md [hunk 2]
```diff
@@ -332,6 +333,7 @@ Contains the request's input data
 |`_id`|`string`| Document unique identifier |
 |`action`|`string`| Kuzzle action to perform |
 |`body`|`object`| Contains request specific data (document content, search queries, ...) |
+| `headers` | `object` | request headers (ie. HTTP headers for HTTP an Websocket protocols) |
```
Comment (scottinet) [inline README.md]: Proposal: `request input headers (e.g. HTTP headers for HTTP or Websocket protocols)`

File: README.md [hunk 3]
```diff
@@ -349,6 +351,7 @@ Other attributes may be defined and will automatically be added to the `args` ob
 | `action` | `string` | `null` | Controller's action to execute |
 | `args` | `object` | *(empty)* | Contains specific request arguments |
 | `body` | `object` | `null` | Request's body (for instance, the content of a document) |
+| `headers` | `object` | request headers (ie. HTTP headers for HTTP an Websocket protocols) |
```
Comment (scottinet) [inline README.md]: same as above

File: README.md (hunk 1 above)
Comment (ballinette) [inline README.md]: @scottinet I just notice that with actual code, we cannot put in `options` arguments a parameter that will be passed to `RequestInput` constructor.

We have two choices :

* keep actual `RequestInput` constructor signature:
```javascript
// request.js
class Request {
  constructor(data, options) {
    this[_input] = new RequestInput(data);
    this[_context] = new RequestContext(options);
  }
}

// Construct a new request:
request = new Request(
  {controller: 'foo', action: 'bar', some: 'args', headers: {x-foo: 'baz'}}, 
  {protocol: 'xxx'}
);
```

* Add `options`, or at least `options.headers` to `RequestInput` constructor:

```javascript
// request.js
class Request {
  constructor(data, options) {
    this[_input] = new RequestInput(data, options.headers);
    this[_context] = new RequestContext(options);
  }
}

// requestInput.js
class RequestInput {
  constructor(data, headers {
// (...)
    this.body = data.body;
    this.headers = headers;
// (...)
  }
}

// Construct a new request:
request = new Request(
  {controller: 'foo', action: 'bar', some: 'args'}, 
  {protocol: 'xxx', headers: {x-foo: 'baz'}}
);
```

I'm divided between these 2 options...
What do you suggest ?

Comment (ballinette) [inline README.md]: 3rd choice:

```javascript
// request.js
class Request {
  constructor(data, options) {
    data.headers ) options.headers
    this[_input] = new RequestInput(data);
    this[_context] = new RequestContext(options);
  }
}

// Construct a new request:
request = new Request(
  {controller: 'foo', action: 'bar', some: 'args'}, 
  {protocol: 'xxx', headers: {x-foo: 'baz'}}
);
```

Comment (scottinet) [inline README.md]: 1. This proposal is a no-go, as there will be conflicts if there is a `headers` optional parameter in a client API route, which might happen considering this is a pretty common term
2. Can be considered
3. Same remark as for the 1st proposal

IMO, either the second option should be considered, or this fourth one, similar to your second proposal without the need to change the constructor of `RequestInput`:

```js
class Request {
  constructor(data, options) {
    this[_input] = new RequestInput(data);

    if (options && options.headers) {
      this[_input].setHeaders(options.headers);
      delete options.headers;
    }

    this[_context] = new RequestContext(options);
  }
}
```

Comment (scottinet) [inline README.md]: Fifth option: change nothing to the current code, and simply add this line _after_ the `Request` object has been created:

`request.input.setHeaders(<headers>);`

File: lib/models/requestInput.js [hunk 4]
```diff
@@ -128,6 +131,13 @@ class RequestInput {
       }
     });
 
+    // initialize `_headers` property after the population of `this.args` attribute
```
Comment (ballinette) [inline lib/models/requestInput.js]: we need to keep a comment here to explain why the following line is not higher in the code.
... but I'm not really satisfied with the wording, so if anyone has a better suggestion...

Comment (xbill82) [inline lib/models/requestInput.js]: looks ok

**Pull request merged**